sklearn/preprocessing/data.py

```diff
@@ -1901,3 +1902,249 @@ def transform(self, X):
+            else:
+                self.col_num_Y_ = len(y[0])
+        else:
+            X = check_array(X)
+            self.col_num_Y_ = 1
+
+        len_data = len(X)
+        self.col_num_X_ = len(X[0])
+        self.count_cache_ = _get_count_dict_3()
+        self.classes_ = [set() for i in range(self.col_num_Y_)]
+
+        inclusion_used = self._get_inclusion_used()
+
+        for inclusion_i in range(len(inclusion_used)):
+            if y is not None:
+                for i in range(len_data):
+                    for j in range(self.col_num_Y_):
+                        X_key = tuple(X[i].take(inclusion_used[inclusion_i]))
+                        if len(y.shape) == 1:
+                            # if y is 1D, y[i] is a scalar; not iterable
+                            y_key = tuple([y[i]])
+                        else:
+                            y_key = tuple(y[i].take([j]))
+                        self.count_cache_[X_key][j][y_key][inclusion_i] += 1
+                        self.classes_[j].add(y_key)
+            else:
+                self.classes_[0].add(0)
+                for i in range(len_data):
+                    X_key = tuple(X[i].take(inclusion_used[inclusion_i]))
+                    self.count_cache_[X_key][0][0][inclusion_i] += 1
+        self.classes_ = [list(enumerate(sorted(ys))) for ys in self.classes_]
+
+        return self
+
+    def transform(self, X):
+        """Transforms X to include the counts learned during 'fit'
+
+        Augments 'X' with a new column containing the counts of each example,
+        conditional on 'y'.
+
+        Parameters
+        ----------
+        X : array
+            The 'X' that we augment with count columns
+            'X' must not be 1 dimensional
+
+        Returns
+        -------
+        transformed : numpy.ndarray
+            The transformed input
+
+        Notes
+        -----
+        The data returned from the transformation will always be a
+        numpy.ndarray
+        """
+
+        check_is_fitted(self, ['count_cache_', 'col_num_X_'])
+
+        X = check_array(X)
+        len_data = len(X)
+        len_classes = 0
+        for ys in self.classes_:
+            len_classes += len(ys)
+
+        num_features = len(X[0])
+        if self.col_num_X_ != num_features:
+            raise ValueError("Dimensions mismatch in X during transform")
+        inclusion_used = self._get_inclusion_used()
+
+        # the number of added cols is the number of unique y vals
+        # multiplied by the number of different inclusion lists
+        num_added_cols = len_classes * len(inclusion_used)
+        transformed = np.zeros((len_data, num_features + num_added_cols))
+        transformed[:, :-num_added_cols] = X
+
+        col_offset_inclusion = 0
+        for inclusion_i in range(len(inclusion_used)):
+            col_offset_y = 0
+            col_offset_inclusion = inclusion_i * len_classes
+            for j in range(self.col_num_Y_):
```

Comment:
If you have arrays instead of single counts in your cache you don't need this loop.